File: superstitioApi/java/superstitioapi/pet/PetManager.java
package superstitioapi.pet;

import basemod.interfaces.PostPowerApplySubscriber;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.MathUtils;
import com.megacrit.cardcrawl.actions.IntentFlashAction;
import com.megacrit.cardcrawl.actions.common.ShowMoveNameAction;
import com.megacrit.cardcrawl.core.AbstractCreature;
import com.megacrit.cardcrawl.core.Settings;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.helpers.TipTracker;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import com.megacrit.cardcrawl.monsters.MonsterGroup;
import com.megacrit.cardcrawl.powers.AbstractPower;
import superstitioapi.InBattleDataManager;
import superstitioapi.SuperstitioApiSubscriber;
import superstitioapi.utils.RenderInBattle;

import static superstitioapi.actions.AutoDoneInstantAction.addToBotAbstract;

public class PetManager implements RenderInBattle, SuperstitioApiSubscriber.AtEndOfPlayerTurnPreCardSubscriber, PostPowerApplySubscriber {
    public final MinionGroup monsterGroup = new MinionGroup(new AbstractMonster[]{});

    public PetManager() {
        RenderInBattle.Register(RenderType.Normal, this);
    }

    public static float calculateSmartDistance(final AbstractCreature m1, final AbstractCreature m2) {
        return (m1.hb_w + m2.hb_w) / 2.0f;
    }

    public static AbstractMonster spawnMonster(final AbstractMonster monsterInstance) {
        final AbstractMonster monster = monsterInstance;
        if (!InBattleDataManager.getPetManager().isPresent()) return monster;
        final MonsterGroup roomMonsters =
                InBattleDataManager.getPetManager().get().monsterGroup;
//        AbstractDungeon.getMonsters();
        float monsterDX = Settings.WIDTH / 2.0f;
        float monsterDY = AbstractDungeon.player.hb.y;
        AbstractMonster lastMonster = null;
        if (!roomMonsters.monsters.isEmpty()) {
            lastMonster = roomMonsters.monsters.get(roomMonsters.monsters.size() - 1);
            monsterDX = lastMonster.drawX;
            monsterDY = lastMonster.drawY;
        }
        if (lastMonster != null)
            monster.drawX = monsterDX - calculateSmartDistance(lastMonster, monster) * Settings.scale;
        else
            monster.drawX = monsterDX - 200.0f * Settings.scale;
        monster.drawY = monsterDY;
        if (monster.drawX < 0.0f || monster.drawX > Gdx.graphics.getWidth()
                || monster.drawY < 0.0f || monster.drawY > Gdx.graphics.getHeight()) {
            monster.drawX = MathUtils.random(0.0f, Gdx.graphics.getWidth());
            monster.drawY = MathUtils.random(Gdx.graphics.getHeight() * 0.15f, Gdx.graphics.getHeight() * 0.85f);
        }
        monster.hb.move(monster.drawX, monster.drawY);
        monster.init();
        monster.applyPowers();
        monster.useUniversalPreBattleAction();
        monster.showHealthBar();
        monster.createIntent();
        monster.usePreBattleAction();
        roomMonsters.add(monster);
        return monster;
    }

    public static AbstractMonster spawnMinion(final Class<? extends AbstractMonster> monsterClass) {
        MinionMonster minionMonster = new MinionMonster(CopyAndSpawnMonsterUtility.motherFuckerWhyIShouldUseThisToCopyMonster(monsterClass));
        return spawnMonster(minionMonster);
    }

    public static AbstractMonster spawnMonster(final Class<? extends AbstractMonster> monsterClass) {
        final AbstractMonster monster =
                CopyAndSpawnMonsterUtility.motherFuckerWhyIShouldUseThisToCopyMonster(monsterClass);
        return spawnMonster(monster);
    }

    public void addPet(AbstractMonster monster) {
        monsterGroup.add(monster);
    }

    @Override
    public void receiveAtEndOfPlayerTurnPreCard() {
        addToBotAbstract(monsterGroup::applyPreTurnLogic);
        monsterGroup.monsters.forEach(this::monsterTurn);
        addToBotAbstract(monsterGroup::applyEndOfTurnPowers);
        addToBotAbstract(monsterGroup::showIntent, 2);
    }

    private void monsterTurn(AbstractMonster monster) {
        if (!monster.isDeadOrEscaped() || monster.halfDead) {
            if (monster.intent != AbstractMonster.Intent.NONE) {
                AbstractDungeon.actionManager.addToBottom(new ShowMoveNameAction(monster));
                AbstractDungeon.actionManager.addToBottom(new IntentFlashAction(monster));
            }

            if (!(Boolean) TipTracker.tips.get("INTENT_TIP") && AbstractDungeon.player.currentBlock == 0 && (monster.intent == AbstractMonster.Intent.ATTACK || monster.intent == AbstractMonster.Intent.ATTACK_DEBUFF || monster.intent == AbstractMonster.Intent.ATTACK_BUFF || monster.intent == AbstractMonster.Intent.ATTACK_DEFEND)) {
                if (AbstractDungeon.floorNum <= 5) {
                    ++TipTracker.blockCounter;
                } else {
                    TipTracker.neverShowAgain("INTENT_TIP");
                }
            }

            monster.takeTurn();
            monster.applyTurnPowers();
        }
    }

    @Override
    public void render(SpriteBatch sb) {
        monsterGroup.render(sb);
    }

    @Override
    public void update() {
        monsterGroup.update();
    }

    @Override
    public void updateAnimation() {
        monsterGroup.updateAnimations();
    }

    @Override
    public void receivePostPowerApplySubscriber(AbstractPower abstractPower, AbstractCreature abstractCreature, AbstractCreature abstractCreature1) {
        monsterGroup.monsters.forEach(AbstractMonster::applyPowers);
    }
}

File: superstitioApi/java/superstitioapi/pet/PetManager.kt
package superstitioapi.pet

import basemod.interfaces.PostPowerApplySubscriber
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.MathUtils
import com.megacrit.cardcrawl.actions.IntentFlashAction
import com.megacrit.cardcrawl.actions.common.ShowMoveNameAction
import com.megacrit.cardcrawl.core.AbstractCreature
import com.megacrit.cardcrawl.core.Settings
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.helpers.TipTracker
import com.megacrit.cardcrawl.monsters.AbstractMonster
import com.megacrit.cardcrawl.monsters.AbstractMonster.Intent
import com.megacrit.cardcrawl.monsters.MonsterGroup
import com.megacrit.cardcrawl.powers.AbstractPower
import superstitioapi.SuperstitioApiSubscriber.AtEndOfPlayerTurnPreCardSubscriber
import superstitioapi.actions.AutoDoneInstantAction
import superstitioapi.renderManager.inBattleManager.InBattleDataManager
import superstitioapi.renderManager.inBattleManager.RenderInBattle
import superstitioapi.utils.ActionUtility.VoidSupplier
import java.util.function.Consumer

class PetManager : RenderInBattle, AtEndOfPlayerTurnPreCardSubscriber, PostPowerApplySubscriber
{
    val monsterGroup: MinionGroup = MinionGroup(arrayOf())

    init
    {
        RenderInBattle.Register(RenderInBattle.RenderType.Normal, this)
    }

    fun addPet(monster: AbstractMonster)
    {
        monsterGroup.add(monster)
    }

    private fun monsterTurn(monster: AbstractMonster)
    {
        if (monster.isDeadOrEscaped && !monster.halfDead)
            return
        if (monster.intent != Intent.NONE)
        {
            AbstractDungeon.actionManager.addToBottom(ShowMoveNameAction(monster))
            AbstractDungeon.actionManager.addToBottom(IntentFlashAction(monster))
        }

        if (!TipTracker.tips["INTENT_TIP"]!! && AbstractDungeon.player.currentBlock == 0 && (monster.intent == Intent.ATTACK || monster.intent == Intent.ATTACK_DEBUFF || monster.intent == Intent.ATTACK_BUFF || monster.intent == Intent.ATTACK_DEFEND))
        {
            if (AbstractDungeon.floorNum <= 5)
            {
                ++TipTracker.blockCounter
            }
            else
            {
                TipTracker.neverShowAgain("INTENT_TIP")
            }
        }

        monster.takeTurn()
        monster.applyTurnPowers()
    }

    override fun receiveAtEndOfPlayerTurnPreCard()
    {
        AutoDoneInstantAction.addToBotAbstract(VoidSupplier(monsterGroup::applyPreTurnLogic))
        monsterGroup.monsters.forEach(Consumer(this::monsterTurn))
        AutoDoneInstantAction.addToBotAbstract(VoidSupplier(monsterGroup::applyEndOfTurnPowers))
        AutoDoneInstantAction.addToBotAbstract(VoidSupplier(monsterGroup::showIntent), 2)
    }

    override fun render(sb: SpriteBatch)
    {
        monsterGroup.render(sb)
    }

    override fun update()
    {
        monsterGroup.update()
    }

    override fun updateAnimation()
    {
        monsterGroup.updateAnimations()
    }

    override fun receivePostPowerApplySubscriber(
        abstractPower: AbstractPower?,
        abstractCreature: AbstractCreature?,
        abstractCreature1: AbstractCreature?
    )
    {
        monsterGroup.monsters.forEach(Consumer(AbstractMonster::applyPowers))
    }

    companion object
    {
        fun calculateSmartDistance(m1: AbstractCreature, m2: AbstractCreature): Float
        {
            return (m1.hb_w + m2.hb_w) / 2.0f
        }

        fun spawnMonster(monsterInstance: AbstractMonster): AbstractMonster
        {
            if (InBattleDataManager.getPetManager() == null)
                return monsterInstance
            val roomMonsters: MonsterGroup =
                InBattleDataManager.getPetManager()!!.monsterGroup
            var monsterDX = Settings.WIDTH / 2.0f
            var monsterDY = AbstractDungeon.player.hb.y
            var lastMonster: AbstractMonster? = null
            if (roomMonsters.monsters.isNotEmpty())
            {
                lastMonster = roomMonsters.monsters[roomMonsters.monsters.size - 1]
                monsterDX = lastMonster.drawX
                monsterDY = lastMonster.drawY
            }
            if (lastMonster != null) monsterInstance.drawX =
                monsterDX - calculateSmartDistance(lastMonster, monsterInstance) * Settings.scale
            else monsterInstance.drawX = monsterDX - 200.0f * Settings.scale
            monsterInstance.drawY = monsterDY
            if (monsterInstance.drawX < 0.0f || monsterInstance.drawX > Gdx.graphics.width || monsterInstance.drawY < 0.0f || monsterInstance.drawY > Gdx.graphics.height)
            {
                monsterInstance.drawX = MathUtils.random(0.0f, Gdx.graphics.width.toFloat())
                monsterInstance.drawY = MathUtils.random(Gdx.graphics.height * 0.15f, Gdx.graphics.height * 0.85f)
            }
            monsterInstance.hb.move(monsterInstance.drawX, monsterInstance.drawY)
            monsterInstance.init()
            monsterInstance.applyPowers()
            monsterInstance.useUniversalPreBattleAction()
            monsterInstance.showHealthBar()
            monsterInstance.createIntent()
            monsterInstance.usePreBattleAction()
            roomMonsters.add(monsterInstance)
            return monsterInstance
        }

        fun spawnMinion(monsterClass: Class<out AbstractMonster>): AbstractMonster
        {
            val minionMonster =
                MinionMonster(CopyAndSpawnMonsterUtility.motherFuckerWhyIShouldUseThisToCopyMonster(monsterClass))
            return spawnMonster(minionMonster)
        }

        fun spawnMonster(monsterClass: Class<out AbstractMonster>): AbstractMonster
        {
            val monster =
                CopyAndSpawnMonsterUtility.motherFuckerWhyIShouldUseThisToCopyMonster(monsterClass)
            return spawnMonster(monster)
        }
    }
}
